ibc-testkit/tests/core/ics02_client/update_client.rs

```diff
@@ -1130,6 +1525,166 @@ fn test_client_update_max_clock_drift() {
 
     let msg_envelope = MsgEnvelope::from(ClientMsg::from(msg));
 
-    let res = validate(&ctx_a, &router_a, msg_envelope);
+    let res = validate(&ctx_a.ibc_store, &router_a, msg_envelope);
     assert!(res.is_err());
 }
+
+/// Builds a `ClientMsg::UpdateClient` for a client with id `client_id` running on the `dest`
+/// context, assuming that the latest header on the source context is `src_header`.
+pub(crate) fn build_client_update_datagram<H: TestHeader, Dst: TestHost>(
+    dest: &MockContext<Dst>,
+    client_id: &ClientId,
+    src_header: &H,
+) -> Result<ClientMsg, RelayerError>
+where
+    HostClientState<Dst>: ClientStateValidation<DefaultIbcStore>,
+{
+    // Check if client for ibc0 on ibc1 has been updated to latest height:
+    // - query client state on destination chain
+    let dest_client_latest_height = dest.light_client_latest_height(client_id);
+
+    if src_header.height() == dest_client_latest_height {
+        return Err(RelayerError::ClientAlreadyUpToDate {
+            client_id: client_id.clone(),
+            source_height: src_header.height(),
+            destination_height: dest_client_latest_height,
+        });
+    };
+
+    if dest_client_latest_height > src_header.height() {
+        return Err(RelayerError::ClientAtHigherHeight {
+            client_id: client_id.clone(),
+            source_height: src_header.height(),
+            destination_height: dest_client_latest_height,
+        });
+    };
+
+    // Client on destination chain can be updated.
+    Ok(ClientMsg::UpdateClient(MsgUpdateClient {
+        client_id: client_id.clone(),
+        client_message: src_header.clone().into(),
+        signer: dummy_account_id(),
+    }))
+}
+
+#[rstest]
+/// Serves to test both ICS-26 `dispatch` & `build_client_update_datagram` functions.
+/// Implements a "ping pong" of client update messages, so that two chains repeatedly
+/// process a client update message and update their height in succession.
```

Comment:
```suggestion
/// Serves to test both ICS-26 `dispatch` & `build_client_update_datagram` functions.
/// Implements a "ping pong" of client update messages, so that two chains repeatedly
/// process a client update message and update their height in succession.
#[rstest]
```